src/libraries/System.Security.Cryptography.Pkcs/src/System/Security/Cryptography/Pkcs/SignerInfo.cs

```diff
@@ -591,100 +587,238 @@ private bool CheckHash(bool compatMode)
-
-            if (_signedAttributes != null)
+        private IncrementalHash GetContentHash(ReadOnlyMemory<byte> content, ReadOnlyMemory<byte>? additionalContent)
+        {
+            IncrementalHash hasher = Helpers.CreateIncrementalHash(GetDigestAlgorithm());
+            if (additionalContent.HasValue)
             {
-                byte[] contentDigest = hasher.GetHashAndReset();
+                hasher.AppendData(additionalContent.Value.Span);
+            }
 
-                AsnWriter writer = new AsnWriter(AsnEncodingRules.DER);
-                {
-                    // Some CMS implementations exist which do not sort the attributes prior to
-                    // generating the signature.  While they are not, technically, validly signed,
-                    // Windows and OpenSSL both support trying in the document order rather than
-                    // a sorted order.  To accomplish this we will build as a SEQUENCE OF, but feed
-                    // the SET OF into the hasher.
-                    if (compatMode)
-                    {
-                        writer.PushSequence();
-                    }
-                    else
-                    {
-                        writer.PushSetOf();
-                    }
+            hasher.AppendData(content.Span);
 
-                    foreach (AttributeAsn attr in _signedAttributes)
-                    {
-                        attr.Encode(writer);
+            return hasher;
+        }
 
-                        // .NET Framework doesn't seem to validate the content type attribute,
-                        // so we won't, either.
+        private static bool VerifyAttributes(
+            ReadOnlySpan<byte> digest,
+            AttributeAsn[] signedAttributes,
+            bool compatMode,
+            bool needsContentAttr,
+            VerifyCallback verify)
+        {
+            bool hasMatchingDigestAttr = false;
+            bool hasContentAttr = false;
 
-                        if (attr.AttrType == Oids.MessageDigest)
-                        {
-                            CryptographicAttributeObject obj = PkcsHelpers.MakeAttribute(attr);
+            AsnWriter writer = new AsnWriter(AsnEncodingRules.DER);
 
-                            if (obj.Values.Count != 1)
-                            {
-                                throw new CryptographicException(SR.Cryptography_BadHashValue);
-                            }
+            // Some CMS implementations exist which do not sort the attributes prior to
+            // generating the signature.  While they are not, technically, validly signed,
+            // Windows and OpenSSL both support trying in the document order rather than
+            // a sorted order.  To accomplish this we will build as a SEQUENCE OF, but feed
+            // the SET OF into the hasher.
+            using (AsnWriter.Scope scope = compatMode ? writer.PushSequence() : writer.PushSetOf())
```

Comment:
Is the ternary forcing the variable to be declared?  It should be fine to do

```suggestion
            using (compatMode ? writer.PushSequence() : writer.PushSetOf())
```

Since you (correctly) never talk about `scope` again.